### PR TITLE
^cooler^ invariance example

### PR DESCRIPTION
**Pull request by cangiuli**
With lots of help from @jonsterling and @ecavallo.

### CODE REVIEW

File: library/cool/invariance.red [hunk 1]
```diff
@@ -41,3 +41,85 @@ def swap-fun/path (A : type) : (f : bool → A) → path _ (swap-fun A (swap-fun
...
+  | suc (_ → xs) → cons triv xs
+  ]
+
+def length (A : type) : list A → nat =
+  elim [
+  | nil → zero
+  | cons _ (_ → n) → suc n
+  ]
+
+-- TODO ??
```
Comment (favonia) [inline library/cool/invariance.red]: what does "TODO" mean here?

Comment (cangiuli) [reply library/cool/invariance.red]: I just wrote it to mean this code might not even be useful? Don't worry about it.

**Comment by cangiuli** [comment]
@favonia @ecavallo So, @jonsterling and I wrote a bunch of really gnarly proofs but we're still stuck and somehow I suspect there's an easier way to do all of this. Does someone want to take a look at how to prove the direct version of `nat->list/is-equiv'`?

File: library/cool/invariance.red [hunk 2]
```diff
@@ -1,9 +1,126 @@
...
+  elim [
+  | zero → zero
+  | suc (_ → ih) → suc (suc ih)
+  ]
+
+def double/list : list unit → list unit =
+  coe 0 1 double/nat in λ i → nat→list/path i → nat→list/path i
+
+def double/list/one : list unit = double/list (cons triv nil)
+meta ⦉ print normalize double/list/one ⦊
```
Comment (jonsterling) [inline library/cool/invariance.red]: btw it should be possible to inline this: `double/list (cons triv nil)` is synthesizable.

Comment (cangiuli) [reply library/cool/invariance.red]: Hm, I get a parse error; what is the syntax? Do I need some brackets?

Comment (jonsterling) [reply library/cool/invariance.red]: Yes, I think you might need parentheses around the expression part.

File: library/cool/nats.red [hunk 3]
```diff
@@ -70,4 +71,78 @@ def nat≃binnat : equiv nat binnat =
...
+  | suc (_ → ih) → not ih
+  ]
+
+def oddq/n≈bn : (n : nat) → path bool (oddq/nat n) (oddq/nat/direct n) =
+  let pf : (n : nat) → path _ (suc/binnat (nat→binnat n)) (nat→binnat (suc n)) =
+    elim [ zero → refl | suc (_ → ih) → refl ]
+  in
+  elim [
+  | zero → refl
+  | suc (n → ih) → λ i → not (trans bool (λ i → oddq (pf n i)) ih i)
```
Comment (cangiuli) [inline library/cool/nats.red]: C{F \\/ E}HM @favonia @ecavallo

Comment (cangiuli) [reply library/cool/nats.red]: According to @favonia this is actually a CJHM @jonsterling

Comment (favonia) [reply library/cool/nats.red]: More accurately, a natural way to finish the proof bumped into something that we haven't spent the time finishing...

Comment (jonsterling) [reply library/cool/nats.red]: Sounds like the unifier, lol.

Comment (favonia) [reply library/cool/nats.red]: See my commit. LOL

**Comment by cangiuli** [comment]
Although it's not fully perfected, I think it might be time to merge this stuff.

**Pull request merged**